File: backend/schemas/projects.py
"""Request/response models for projects and project membership."""
from typing import List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from backend.schemas.common import Page, UtcDatetime

ProjectEnvironment = Literal["development", "staging", "production"]
ProjectCriticality = Literal["low", "medium", "high", "critical"]
ProjectStatus = Literal["active", "archived"]
ProjectRole = Literal["owner", "security", "developer", "viewer"]


class Technology(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    version: str = Field(default="", max_length=50)


class ProjectCreate(BaseModel):
    workspace_id: UUID
    name: str = Field(min_length=1, max_length=200, examples=["Customer Portal"])
    domain: Optional[str] = Field(default=None, max_length=255, examples=["portal.acme.com"])
    environment: ProjectEnvironment = Field(examples=["production"])
    criticality: ProjectCriticality = Field(examples=["high"])
    internet_facing: bool = Field(default=False)
    technologies: List[Technology] = Field(default_factory=list, max_length=100)


class ProjectUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    domain: Optional[str] = Field(default=None, max_length=255)
    environment: Optional[ProjectEnvironment] = None
    criticality: Optional[ProjectCriticality] = None
    internet_facing: Optional[bool] = None
    technologies: Optional[List[Technology]] = Field(default=None, max_length=100)


class ProjectItem(BaseModel):
    id: UUID = Field(examples=["7c9e6679-7425-40de-944b-e07fc1f90ae7"])
    workspace_id: UUID
    name: str
    domain: Optional[str] = None
    environment: ProjectEnvironment
    criticality: ProjectCriticality
    internet_facing: bool
    technologies: List[Technology]
    status: ProjectStatus
    archived_at: Optional[UtcDatetime] = None
    owner_user_id: UUID
    created_at: UtcDatetime
    updated_at: UtcDatetime


class ProjectPage(Page[ProjectItem]):
    pass


class ProjectMemberAdd(BaseModel):
    user_id: UUID
    project_role: ProjectRole


class ProjectMemberRoleChange(BaseModel):
    project_role: ProjectRole


class ProjectMemberItem(BaseModel):
    id: UUID
    project_id: UUID
    user_id: UUID
    project_role: ProjectRole
    created_at: UtcDatetime
    updated_at: UtcDatetime


class ProjectMemberList(BaseModel):
    items: List[ProjectMemberItem]

File: frontend/src/lib/api/projects.ts
import { apiDelete, apiGet, apiPatch, apiPost } from './client';
import type { Page } from './chatbot';

export type ProjectEnvironment = 'development' | 'staging' | 'production';
export type ProjectCriticality = 'low' | 'medium' | 'high' | 'critical';
export type ProjectStatus = 'active' | 'archived';
export type ProjectRole = 'owner' | 'security' | 'developer' | 'viewer';

export interface Technology {
  name: string;
  version: string;
}

export interface Project {
  id: string;
  workspace_id: string;
  name: string;
  domain: string | null;
  environment: ProjectEnvironment;
  criticality: ProjectCriticality;
  internet_facing: boolean;
  technologies: Technology[];
  status: ProjectStatus;
  archived_at: string | null;
  owner_user_id: string;
  created_at: string;
  updated_at: string;
}

export interface ProjectMember {
  id: string;
  project_id: string;
  user_id: string;
  project_role: ProjectRole;
  created_at: string;
  updated_at: string;
}

export interface ProjectCreateInput {
  workspace_id: string;
  name: string;
  domain?: string | null;
  environment: ProjectEnvironment;
  criticality: ProjectCriticality;
  internet_facing?: boolean;
  technologies?: Technology[];
}

export interface ProjectUpdateInput {
  name?: string;
  domain?: string | null;
  environment?: ProjectEnvironment;
  criticality?: ProjectCriticality;
  internet_facing?: boolean;
  technologies?: Technology[];
}

export interface ProjectFilters {
  page?: number;
  pageSize?: number;
  workspaceId?: string;
  includeArchived?: boolean;
}

export function listProjects(filters: ProjectFilters = {}): Promise<Page<Project>> {
  const params = new URLSearchParams();
  params.set('page', String(filters.page ?? 1));
  params.set('page_size', String(filters.pageSize ?? 20));
  if (filters.workspaceId) params.set('workspace_id', filters.workspaceId);
  if (filters.includeArchived) params.set('include_archived', 'true');
  return apiGet<Page<Project>>(`/api/projects?${params.toString()}`);
}

export function getProject(id: string): Promise<Project> {
  return apiGet<Project>(`/api/projects/${id}`);
}

export function createProject(input: ProjectCreateInput): Promise<Project> {
  return apiPost<Project>('/api/projects', input);
}

export function updateProject(id: string, input: ProjectUpdateInput): Promise<Project> {
  return apiPatch<Project>(`/api/projects/${id}`, input);
}

export function archiveProject(id: string): Promise<Project> {
  return apiPost<Project>(`/api/projects/${id}/archive`);
}

export function listProjectMembers(projectId: string): Promise<{ items: ProjectMember[] }> {
  return apiGet<{ items: ProjectMember[] }>(`/api/projects/${projectId}/members`);
}

export function addProjectMember(
  projectId: string,
  userId: string,
  projectRole: ProjectRole,
): Promise<ProjectMember> {
  return apiPost<ProjectMember>(`/api/projects/${projectId}/members`, {
    user_id: userId,
    project_role: projectRole,
  });
}

export function changeProjectMemberRole(
  projectId: string,
  userId: string,
  projectRole: ProjectRole,
): Promise<ProjectMember> {
  return apiPatch<ProjectMember>(`/api/projects/${projectId}/members/${userId}`, {
    project_role: projectRole,
  });
}

export function removeProjectMember(projectId: string, userId: string): Promise<void> {
  return apiDelete<void>(`/api/projects/${projectId}/members/${userId}`);
}
